WordPress/Jetpack/Classes/NUX/New Landing Screen/Views/InfiniteScrollerView.swift

```diff
@@ -0,0 +1,114 @@
+        guard let stackView = stackView, !stackView.arrangedSubviews.isEmpty else {
+            return nil
+        }
+
+        return stackView.arrangedSubviews[0].frame.size.height
+    }
+
+    /// Initializes a view that produces an infinite scrolling effect.
+    /// - Parameter viewBuilder: Closure that generates a `UIView`. Repeatedly called to fill a `UIStackView` that will be scrolled.
+    convenience init(_ viewBuilder: @escaping ((Int) -> UIView)) {
+        self.init()
+        self.viewBuilder = viewBuilder
+        displayLink = CADisplayLink(target: self, selector: #selector(step))
+
+        setupScrollView()
+        let stackView = setupStackView()
+        addSubview(stackView)
+
+        stackView.topAnchor.constraint(equalTo: topAnchor).isActive = true
+        stackView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
+        stackView.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
+        stackView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
+        stackView.widthAnchor.constraint(equalTo: widthAnchor).isActive = true
+
+        self.stackView = stackView
+        addViewsToStackView()
+
+        displayLink?.add(to: .current, forMode: .default)
+    }
+
+    private func addViewsToStackView() {
+        guard let viewBuilder = viewBuilder else { return }
+        guard let stackView = stackView else { return }
+
+        /// Note: This class makes the assumption that the view generated by `viewBuilder()` is taller than
+        /// the scroll view's frame. To support views shorter than the frame it'd need to calculate how many times
+        /// to repeat it based on the built view's height.
+        ///
+        /// Currently, five views are generated because of an even / odd congruency requirement by the JetpackLandingScreen.
+        for i in 0...4 {
+            stackView.addArrangedSubview(viewBuilder(i))
+        }
+    }
+
+    private func setupScrollView() {
+        delegate = self
+        bounces = false
+        showsVerticalScrollIndicator = false
+        showsHorizontalScrollIndicator = false
+        isUserInteractionEnabled = false
+    }
+
+    private func setupStackView() -> UIStackView {
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.distribution = .fillEqually
+        stackView.spacing = 0
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+
+        return stackView
+    }
+
+    /// Called on each `CADisplayLink` frame and sets the vertical content offset if a scroller delegate is set.
+    @objc private func step() {
+        guard let scrollerDelegate = self.scrollerDelegate else { return }
+        guard let singleViewHeight = self.singleViewHeight else { return }
+        // the rate shouldn't be higher than the size of the view
+        guard scrollerDelegate.rate < singleViewHeight else { return }
+
+        contentOffset.y += scrollerDelegate.rate
```

Comment:
I tested the animation on a 120Hz ( iPad Pro ) and a 60Hz Device ( iPhone 14 Pro Simulator ) and I noticed that the scrolling is slightly faster on 120Hz devices. 

This happens because on 120Hz devices, the `step` method is called 120 times per 1 second, or every 1/120 ( = 0.008 ) seconds. Assuming the rate is -1, this means the scrollView will scroll by -120pts per 1 second.

For 60Hz devices, the `step` is called only 60 times per second or every 1/60 ( = 0,016 ) seconds. This means the scrollView will scroll by -60pts per 1 second.

And I think what we want instead is for the scrollView to scroll at the same speed no matter the device refresh rate.

I added a code that logs the distance scrolled every time `step` is called. Or in other words, it logs the distance scrolled for every frame. Here is the result:

| 120Hz Device | 60Hz Device |
| -------------- | ------------- |
| The rate is -1 and the scrollView scrolled by -120pts in 1 second. ![](https://user-images.githubusercontent.com/9609223/194861534-7f99763b-8086-464c-8f5a-c36aad6464ea.png) | The rate is -1 and the scrollView scrolled by -60pts in 1 second. ![](https://user-images.githubusercontent.com/9609223/194861602-4bdb5d90-a71e-4b84-b466-0eda2ef519ae.png) |

This was an interesting problem to think about so I tried to look for a solution. The idea is to define a scrolling `velocity` and we calculate the `rate` internally depending on the device refresh rate, like this:

```swift
// Represents the scrolled distance per second. The delegate should provide this value.
let velocity: Double = -60 // -60pts/s 

// For low velocities, like 10pts/s, the rate is 10 * (1/120) = 0.08 for 120Hz devices. 
// And there is a UIKit limitation where we can only add  ( or subtract ) at least 0,33pts depending on the screen scale. 
// Assuming `self.contentOffset.y` is 0, adding 0.08 does nothing, the contentOffset remains 0.
private lazy var newContentOffset = contentOffset

@objc private func step(displayLink: CADisplayLink) {
    let rate = velocity * displayLink.duration
    newContentOffset.y += rate
    contentOffset = newContentOffset
}
```

I logged the distance and velocity remained the same for both 120Hz and 60Hz devices:

| 120Hz Device | 60Hz Device |
| -------------- | ------------- |
| ![ipad-velocity-approach](https://user-images.githubusercontent.com/9609223/194866442-c587413f-20ca-45d5-80da-e95856b98db4.png) | ![iphone-velocity-approach](https://user-images.githubusercontent.com/9609223/194866404-e571e277-56d8-45d2-8c25-1fe185983324.png) |

Reply:
I think this issue is not blocking, we can merge this PR and address the issue in another PR. Also I'm happy to open a PR with the changes I explained above. Let me know what you think. 🙏

Reply:
Good catch @salimbraksa! We definitely should decouple the frames-per-second for the device from the scroll velocity. So, the delegate should respond with "The current rate (some distance over time) is this value." and the infinite scroll view should do its best to achieve it with the data we get from the `CADisplayLink`.

What you have looks good. I'm going to give it a look tomorrow and see what we can do in this PR. Thanks!

Reply:
👋 @salimbraksa - I added 8046bf80e2aa3df072133db926f67d6e5782d193 to account for this and will be interested in your thoughts.